google/ggrc-core
# GGRC-6416/GGRC-5110 Improve performance for assessment generation with template and CADs

# Dependencies

This PR is `on hold` until:
- [x] Expected timings result are provided;
- [x] Verifiers propagation for created assessment is clarified with BA.
- [x] #9759 

# Issue description

Assessment generation from template with CADs should work faster.

# Steps to test the changes

1. Create CADs of all types for assessment objects (5 at the time: _text_, _rich text_, _date_, _checkbox_, _dropdown_);
2. Create a new program;
3. Map 100 controls to program;
4. Create an audit;
5. Go to audit page;
6. Create an assessment template with all types of local CADs (6 at the time: _text_, _rich text_, _date_, _checkbox_, _dropdown_, _person_);
7. Generate assessments using created template and snapshots of mapped controls (100 assessments should be generated).

# Solution description

Following changes where made:

* Revision relationship on snapshot was changed to load snapshots along with revisions in order to reduce number of DB queries.

* Snapshot's `eager_query` now loads revisions with event's `action` field since it is needed for revision description. Revision's `eager_query` was also changed to load event's `action`.

* `deferred` was removed from `Snapshot` and `Person` fields to reduce number of DB queries and remove `_is_deferred_loaded` call during ACL serialization.

* `handle_assessment_post` is completely rewritten to work more effectively when several assessments being generated at the same time (before it was designed to process assessments one by one).



# Sanity checklist




# PR Review checklist
